File: StringKata/StringCalculator.java
import java.io.*;
import java.util.ArrayList;

class StringCalculator{

	public int counter = 0;

	public int GetCalledCount(){
		return counter;
	}

	public int Add(String numbers) throws Exception{
		counter++;

		if (numbers.isEmpty()== true){
			return 0;
		}

		else{
			int sum=0;
			int indexFirstNumber=0;//indice prima cifra del numero
			ArrayList<String> negativeNumbers = new ArrayList<String>();

			//Controllo se c'è un altro delimitatore da supportare oltre la virgola e \n
			char delimitatore='\0';
			if(numbers.substring(0,2).equals("//")){
				delimitatore = numbers.charAt(2);
				indexFirstNumber=4;
			}

			for(int i=indexFirstNumber; i<numbers.length(); i++){
				if (numbers.charAt(i)==',' || numbers.charAt(i)=='\n' || numbers.charAt(i)==delimitatore){
					String numberToSum=numbers.substring(indexFirstNumber,i);

					//Verifico se è un numero negativo, se non lo è sommo
					if(numberToSum.indexOf("-")==-1){
						sum+=Integer.parseInt(numberToSum);
						indexFirstNumber= i+1;
					}
					//Se numero negativo aggiungo alla lista
					else{
						negativeNumbers.add(numberToSum);
						indexFirstNumber= i+1;
					}
				}
			}

			//ULTIMO NUMERO
			String numberToSum=numbers.substring(indexFirstNumber);
			if(numberToSum.indexOf("-")==-1){
				sum+=Integer.parseInt(numberToSum);
			}
			else{
				negativeNumbers.add(numberToSum);
			}

			//Ecezzione se numeri negativi altrimenti restituisco la somma
			if(!negativeNumbers.isEmpty()){
				throw new Exception("Numeri negativi non permessi, presenti: "+ negativeNumbers);
			}
			return sum;
		}

	}
}

File: StringKata2/src/main/kotlin/StringCalculator.kt
class StringCalculator {
    private var counter: Int =0

    fun add(stringa: String): Int {
        counter++

        if (stringa.isEmpty()) {
            return 0
        }
        val numeri = numeri(stringa)

        val numeriNegativi= numeriNegativi(numeri)

        if (numeriNegativiPresenti(numeriNegativi)) {
            throw Exception("Numeri negativi non permessi, presenti %s".format(numeriNegativi.toString()))
        }

        return numeri.sum();
    }

    private fun numeri(stringa: String): List<Int> {
        val delimiter1 = ","
        val delimiter2 = "\n"

        return estraiStringaPulita(stringa).split(delimiter1, delimiter2, estraiDelimitatore(stringa))
            .map { it.toInt() }
            .filter { it <= 1000 }
    }

    private fun estraiStringaPulita(stringa: String): String {
        val stringaPulita: String = if (stringa.startsWith("//[")) {
            stringa.substring(stringa.indexOf("\n") + 1)
        } else if (stringa.startsWith("//")) {
            stringa.substring(4)
        } else {
            stringa
        }
        return stringaPulita
    }

    private fun estraiDelimitatore(stringa: String): String {
        val delimiter3 = if (stringa.startsWith("//[")) {
            stringa.substring(3, stringa.indexOf('\n') - 1)
        } else if (stringa.startsWith("//")) {
            stringa[2].toString()
        } else {
            " "
        }
        return delimiter3
    }

    private fun numeriNegativi(numeri: List<Int>): List<Int> {
        return numeri.filter{ it<0}
    }

    private fun numeriNegativiPresenti(numeriNegativi: List<Int>) =
        numeriNegativi.isNotEmpty()

    fun getCalledCount(): Int {
        return counter
    }



}